### PR TITLE
Allow adding rows with null metadata

### LINKED ISSUE #1256: tables do not allow setting of metadata=None even if 'null' is allowed for metadata
We made it so that we can allow empty metadata by setting `type = ['object', 'null']` in the metadata schema. If so, then
```
schema.validate_and_encode_row(None)
```
returns `b''`. However, you *can't* actually do
```
table.add_row(metadata=None)
```
because you get the error
```
E           tskit.exceptions.MetadataValidationError:
... 
E           
E           On instance:
E               {}

```

The culprit is right here: https://github.com/tskit-dev/tskit/blob/main/python/tskit/metadata.py#L571
because `metadata=None` gets turned into `{}` here: https://github.com/tskit-dev/tskit/blob/main/python/tskit/tables.py#L549

A test and possible solution to follow.

### PR DESCRIPTION
**Pull request by petrelharp**
Closes #1256 although probably not in the best possible way.

### CODE REVIEW

**Comment by petrelharp** [comment]
*Edit:* This is *definitely* not the best way to do it.

**Review by jeromekelleher — COMMENTED** [review]
LGTM, but I'll defer to @benjeffery

File: python/tests/test_metadata.py [hunk 1]
```diff
@@ -440,15 +440,23 @@ def test_bad_top_level_type(self):
...
-        schema = {
-            "codec": "json",
-            "type": ["object", "null"],
-            "properties": {"one": {"type": "string"}, "two": {"type": "number"}},
-        }
-        ms = metadata.MetadataSchema(schema)
-        row_data = {"one": "tree", "two": 5}
-        assert ms.decode_row(ms.validate_and_encode_row(row_data)) == row_data
-        assert ms.decode_row(ms.validate_and_encode_row(None)) is None
+        for codec in ["json", "struct"]:
```
Comment (petrelharp) [inline python/tests/test_metadata.py:443]: This was a test I added at first before I figured out where the problem was, and figured it's a good one, so I'd leave it in.

File: python/tests/test_tables.py [hunk 2]
```diff
@@ -851,6 +851,32 @@ def test_round_trip_set_columns(self):
...
+                "title": "Example Metadata",
+                "type": ["object", "null"],
+                "properties": {
+                    "a": {"type": "number", "binaryFormat": "i"},
+                },
+                "required": ["a"],
+                "additionalProperties": False,
+            },
+        )
+        metadata_list = [{"a": 4}, None, {"a": 5}]
```
Comment (benjeffery) [inline python/tests/test_tables.py:868]: Might as well add `{}` here too to check it doesn't become `None` with future fiddling.

**Review by benjeffery — APPROVED** [review]
What misgivings did you have? This seems like the right way to fix this!
One test tweak suggestion.

File: python/tests/test_tables.py (hunk 2 above)
Comment (petrelharp) [reply python/tests/test_tables.py:868]: That won't validate on this schema, but I'll add another test.

File: python/tests/test_tables.py [hunk 3]
```diff
@@ -851,6 +851,61 @@ def test_round_trip_set_columns(self):
...
+                        **self.input_data_for_add_row(),
+                        "metadata": md,
+                    }
+                )
+            assert table.num_rows == len(metadata_list)
+            for md, row in zip(metadata_list, table):
+                assert md == row.metadata
+
+    @pytest.mark.skip("Fails, but shouldn't?")
+    def test_set_with_optional_properties(self):
```
Comment (petrelharp) [inline python/tests/test_tables.py:882]: > Might as well add {} here too to check it doesn't become None with future fiddling.

Good suggestion, @benjeffery! This fails for totally different reasons on both codecs.

Comment (petrelharp) [reply python/tests/test_tables.py:882]: With `struct` it fails with:
```
tskit/metadata.py:452: in <lambda>
    else b"".join(
>           sub_encoder(obj[key]) for key, sub_encoder in sub_encoders.items()
        )
    )
E   KeyError: 'a'
```
and with `json` it fails by turning the None into `{}`:
```
>               assert md == row.metadata
E               assert None == {}
E                +  where {} = IndividualTableRow(flags=3, location=array([], dtype=float64), parents=array([], dtype=int32), metadata={}).metadata
```

File: python/tests/test_tables.py [hunk 4]
```diff
@@ -851,6 +851,67 @@ def test_round_trip_set_columns(self):
...
+                table.add_row(
+                    **{
+                        **self.input_data_for_add_row(),
+                        "metadata": md,
+                    }
+                )
+            assert table.num_rows == len(metadata_list)
+            for md, row in zip(metadata_list, table):
+                if isinstance(md, dict):
+                    # struct inserts defaults, json does not
```
Comment (petrelharp) [inline python/tests/test_tables.py:908]: This was surprising. Does the JSON codec do anything with defaults?

File: python/tskit/metadata.py [hunk 5]
```diff
@@ -446,13 +446,23 @@ def make_object_or_null_encode(cls, sub_schema):
             key: StructCodec.make_encode(prop)
             for key, prop in sub_schema["properties"].items()
         }
-        return (
-            lambda obj: b""
-            if obj is None
-            else b"".join(
-                sub_encoder(obj[key]) for key, sub_encoder in sub_encoders.items()
-            )
-        )
```
Comment (petrelharp) [inline python/tskit/metadata.py:455]: This was a bug, I think - `make_object_encode` was using defaults, but `make_object_or_null_encode` was not, leading to a `KeyError` when trying to encode objects without all the required properties.

File: python/tskit/metadata.py [hunk 6]
```diff
@@ -568,7 +578,10 @@ def __init__(self, schema: Optional[Mapping[str, Any]]) -> None:
             self._validate_row = TSKITMetadataSchemaValidator(schema).validate
             self.encode_row = codec_instance.encode
             self.decode_row = codec_instance.decode
-            self.empty_value = {}
+            if "type" in schema and "null" in schema["type"]:
```
Comment (petrelharp) [inline python/tskit/metadata.py:581]: Before I had this only putting the empty value as None if there were required attributes. I think this is wrong, and should be this way - None is "empty" if allowed.

**Comment by petrelharp** [comment]
I think this is ready to go now, although maybe "json doesn't use defaults" is a bug for another PR?

File: python/tests/test_tables.py (hunk 4 above)
Comment (benjeffery) [reply python/tests/test_tables.py:907]: It's in work here: https://github.com/tskit-dev/tskit/pull/1149

File: python/tests/test_metadata.py (hunk 1 above)
Comment (jeromekelleher) [reply python/tests/test_metadata.py:443]: Good to pytest.mark.parametrize this rather than looping

File: python/tests/test_tables.py [hunk 7]
```diff
@@ -851,6 +851,66 @@ def test_round_trip_set_columns(self):
                 assert table[j].metadata == metadata_column[j]
                 assert table[j + num_rows].metadata == metadata_column[j]
 
+    def test_set_null_metadata(self):
+        for codec in ["struct", "json"]:
```
Comment (jeromekelleher) [inline python/tests/test_tables.py:855]: Likewise here:
```
@pytest.mark.parametrize("codec", ["struct", "json"])
def test_set_null_metadata(self, codec):
     table = self.table_class() 
     # ect
```

It's a great pattern once you get used to using it!

**Review by jeromekelleher — COMMENTED** [review]
LGTM @petrelharp, minor comments.

**Comment by benjeffery** [comment]
@petrelharp I'm adding some commits here, I'll also address the comments.

**Comment by benjeffery** [comment]
Ok, I've added a skip to the test that needs JSON default handling. If you're happy @petrelharp I'll squash and merge.

File: python/tests/test_tables.py [hunk 8]
```diff
@@ -851,65 +851,75 @@ def test_round_trip_set_columns(self):
...
-                {
-                    "codec": f"{codec}",
-                    "title": "Example Metadata",
-                    "type": ["object", "null"],
-                    "properties": {
-                        "a": {"type": "number", "binaryFormat": "i"},
-                    },
-                    "required": ["a"],
-                    "additionalProperties": False,
+    # Properties are not optional for struct
```
Comment (petrelharp) [inline python/tests/test_tables.py:854]: Is this because the default for struct is to require properties, and not for json?

Comment (benjeffery) [reply python/tests/test_tables.py:854]: Yes, struct has no way to encode missingness in the current design.

Comment (petrelharp) [reply python/tests/test_tables.py]: Ah, never mind, I got it.

**Comment by petrelharp** [comment]
I tweaked your tweak to separate out some of the things we're testing and make it so we test with an explicit "required" entry; go ahead and squash & merge if you like it? Or, tweak more?

**Comment by benjeffery** [comment]
Thanks for clearing this up @petrelharp! Accommodating legacy metadata has lead to some wrinkles in the design, but I think we're getting there.

**Pull request merged**